denonavr/denonavr.py

```diff
@@ -1212,6 +1263,70 @@ def set_input_func(self, input_func):
                           input_func)
             return False
 
+    @sound_mode.setter
+    def sound_mode(self, sound_mode):
+        """Setter function for sound_mode to switch sound_mode of device."""
+        self.set_sound_mode(sound_mode)
+
+    def set_sound_mode(self, sound_mode):
+        """
+        Set sound_mode of device.
+
+        Valid values depend on the device and should be taken from
+        "sound_mode_list".
+        Return "True" on success and "False" on fail.
+        """
+        # For selection of sound mode other names then at receiving sound modes
+        # have to be used
+        # Therefore source mapping is needed to get sound_mode
+        # Create command URL and send command via HTTP GET
+        command_url = self._urls.command_sel_sound_mode + sound_mode
+        # sent command
+        try:
+            if self.send_get_command(command_url):
+                self._sound_mode = sound_mode
+                return True
+            else:
+                return False
+        except requests.exceptions.RequestException:
+            _LOGGER.error("Connection error: sound mode function %s not set.",
+                          sound_mode)
+            return False
+
+    def set_sound_mode_dict(self, sound_mode_dict):
+        Error_msg = ("Syntax of sound mode dictionary not valid, "
+                     "use: OrderedDict([('COMMAND', ['VALUE1','VALUE2'])])")
+        if type(sound_mode_dict) == OrderedDict:
+            mode_list = list(sound_mode_dict.values())
+            for sublist in mode_list:
+                if type(sublist) == list:
+                    for element in sublist:
+                        if type(element) != str:
+                            _LOGGER.error(Error_msg)
+                            return False
+                else:
+                    _LOGGER.error(Error_msg)
+                    return False
+        else:
+            _LOGGER.error(Error_msg)
+            return False
+        self._sound_mode_dict = sound_mode_dict
+        return True
+
+    def match_sound_mode(self, sound_mode_raw):
+        try:
+            mode_list = list(self._sound_mode_dict.values())
```

Comment:
When I look at these lines, I get the impression you created the `SOUND_MODE_MAPPING` and `self._sound_mode_dict` the wrong way and mixed key and values.
You are using it to get the sound modes from the raw values. For me this is a normal dictionary with each  `sound_mode_raw` as key and its` sound_mode` as value. That would simplify this part here a lot.
`sound_mode_list` would be a set `set(self._sound_mode_dict.values())` then.

Reply:
In principle you would be right, and it would be nicer to do it that way if each sound_mode_raw would correspond to exactly 1 matched sound mode.
Unfortunately this is not the case, and multiple sound_mode_raw's can be matched to the same matched sound_mode.

I did think about this before I set it up, and in my opinion it is more elegant to do it like I did it because now I have a single matched sound_mode as key, and then all corresponding raw_sound_modes as list in the value corresponding to that key. In this way I do not need to repeat the matched sound_mode as the value for several raw_sound_mode keys (if you would do it the other way around).

Therefore my method seemed more compact and in my opinion more elegant for this situation.

Reply:
In principle I chose for keeping the self._sound_mode_dict as compact and most eligant as I could make it, and having a bit more complicated code in the match_sound_mode function.

I did this because users of this library and Home Assistant will probably want to tweak the self._sound_mode_dict to have the exact mapping values for their model of receiver in their if it is not yet incorperated in the default mapping dict.
However those users will in principle never have to deal with the code of the match_sound_mode function, so it is no problem if that gets a bit more complicated.

Reply:
It does not look too efficient to replace a dictionary key access by a nested loop across all values of a dictionary during run time.

Maintaining own mapping rules is indeed easier using your way of modelling.

Thus, I would use a dictionary with key: `sound_mode_raw` and value: `sound_mode` during run time as I suggested.
This dictionary could be created by the `set_sound_mode_dict` method you already created from a dictionary with the structure you suggested (key: `sound_mode`, value: `list of sound_mode_raw`)

Reply:
@scarface-4711 smart, it is indead more efficient to use a simple dict key acces during runtime, I have implemented this change in the code.